pkg/trace/sampler/remote_rates.go

```diff
@@ -0,0 +1,218 @@
+package sampler
+
+import (
+	"sync"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/config/remote/service"
+	"github.com/DataDog/datadog-agent/pkg/proto/pbgo"
+	"github.com/DataDog/datadog-agent/pkg/trace/config"
+	"github.com/DataDog/datadog-agent/pkg/trace/pb"
+	"github.com/DataDog/datadog-agent/pkg/trace/watchdog"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+// RemoteRates sharded by signature. Allowing an independant feedback
+// loop per signature. RemoteRates adjusts sampling rates communicated
+// to the agent based on the observed traffic. It targets a TPS configured
+// remotely.
+type RemoteRates struct {
+	samplers   map[Signature]*Sampler
+	tpsTargets map[Signature]float64
+	mu         sync.RWMutex // protects concurrent access to samplers and tpsTargets
+
+	exit    chan struct{}
+	stopped chan struct{}
+}
+
+func newRemoteRates(conf *config.AgentConfig) *RemoteRates {
+	if !conf.EnabledRemoteRates {
+		return nil
+	}
```

Comment:
I think you could still return a non-nil value here, even if it's disabled. You could instead check inside each exported method and simply return, obtaining basically a no-op. This way, you will be able to drop all the `nil` comparisons inside `PrioritySampler` and the code will become cleaner as a result.

Reply:
Circling back on this, it also looks like you have a lot of exported methods, so possibly not worth it. I'll let you decide.